async/timeout.ts

```diff
@@ -0,0 +1,18 @@
+import { deferred } from "./deferred.ts";
+
+export class TimeoutError extends Error {
+  constructor() {
+    super("Timeout");
+    this.name = "TimeoutError";
+  }
+}
+
+/**
+ * Create a promise which will be rejected with TimeoutError when a given delay is exceeded.
+ */
+export function timeout<T>(p: Promise<T>, delay: number): Promise<T> {
```

Comment:
I don't really like the name, I think it's too close to `setTimeout` and I expect would cause confusion. Maybe `race` or `deadline` would be a better name for this function.

Reply:
OK. I'll use `deadline` 👍